### PR TITLE
Refactor web provider

### PR DESCRIPTION
**Pull request by dleangen**
Hi,

I would like to refactor the web provider in order to make it more flexible and more readable.

The idea is to separate the different "parts" into configurable ConditionalServlets. This will provide more flexibility, as each servlet can be either configured, or easily replaced by implementing a different ConditionalServlet and updating the service.ranking.

As the first step, I have attempted to pull out the Cache functionality. This is just to get the dialogue started. Nothing is actually using it yet.

If you agree with what I'm doing so far, then I will continue step by step. Please let me know.

### CODE REVIEW

**Comment by dleangen** [comment]
Just tried integrated the extracted Cache into the code. The result is very ugly. :-(

It is my hope that this will get cleaned up significantly along the way.

**Comment by dleangen** [comment]
Made a few more changes. I am slowly making my way towards my vision.

Please feel free to add comments at any time. :-)

**Comment by pkriens** [comment]
In general I am not a fan of polishing code, especially if it creates additional classes only for purity sake. So why extract the cache? Is there a reasonable use case for another type of cache?

Then again, it is not my time :-) So I am ok but I think there are more interesting areas to improve?

**Comment by dleangen** [comment]
Fair comment. And I agree, btw.

In my case, central to my application is the ability to ship static web resources from bundles. That part of the WebServer is what I am most interested, but I'm not able to use it the way I want.

WebServer does a lot of stuff (from memory: static files, webresources, plugin contributions...), and I am trying to separate out that stuff so I can have more control and flexibility. The cache seems to be common across all of those functions, and in any case, seemed to be a good starting point. To be honest, I have been having trouble wrapping my head around all that the class does, so I have been trying to chip away at it. The Cache was the first place I thought I understood, and my understanding is increasing each time I touch the code. I am starting to see the final picture. Just a matter of hacking my way there.

**Comment by pkriens** [comment]
I do not want to hold you back. I like that you're active in it. I will accept the PR once you feel happy and I feel confident it does not break others. So don't let me stop you.

**Comment by dleangen** [comment]
Thanks Peter. I'll let you know.

**Comment by pkriens** [comment]
How are we on this?

**Comment by dleangen** [comment]
Hi Peter,

Very close. Hope to finish today, but expect to finish by this weekend at the latest. Wasn't expecting it would drag out so long. Thanks for your patience!

**Comment by dleangen** [comment]
I just rebased to HEAD, and now my workspace is broken.

What do I need to do to get back in business? Or should I move back to an older commit and work from there?

**Comment by pkriens** [comment]
Broken in what way? You had conflicts? Can't you just fix the conflicts?

**Comment by dleangen** [comment]
Hmmm. It's working again now. I had just assumed it was related to the recent changes in HEAD, and wondered if I needed to make some change.

Either you are playing mind games with me, or I am playing mind games with myself. ;-P

Anyway, back in business...

**Comment by dleangen** [comment]
Ok, I think I'm about done. Here are the highlights:

 * Code is (IMO) a more readable
 * Documentation available in the readme file, which explains what I did
 * Added tests cases to also show the intent
 * Much more configurable
 * Unused stuff clipped out

The previous code was very dense. There was a lot of stuff going on, and (and least for me) I had a lot of trouble understanding it.

In order to make the code more configurable, I first needed to understand it. To do that, I had to try to untangle it. The current style is quite different. I hope that's ok. Yes, there are more classes than before, but it is all with the goal of configurability and readability. I hope I was able to accomplish that goal in the eyes of people other than myself. :-)

I think the test cases show that it is working as intended, but I guess the real test will come once it's tried out against "legacy" code. I believe it should work, though.

Looking forward to comments.

**Comment by dleangen** [comment]
(Friendly) ping. :-)

**Comment by pkriens** [comment]
Friendly ping back … I am a bit too overload this week … Hope to get there next week. :-(

Kind regards,

	Peter Kriens

**Comment by dleangen** [comment]
Just pinging again to make sure this stays in the list of things todo. :-)

**Comment by njbartlett** [comment]
I'd also like to see this merged.

Though I haven't actually reviewed the PR content, sorry.

**Comment by pkriens** [comment]
I've not forgotten you but it is just so darned busy at the moment :-(

**Comment by dleangen** [comment]
No worries, I completely understanding. Thanks for letting me know. :-)

File: osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/EnrouteApplicationIndexServer.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+				"name=" + EnrouteApplicationIndexServer.NAME, 
+		}, 
+		name = EnrouteApplicationIndexServer.NAME, 
+		configurationPid = BundleMixinServer.NAME,
+		configurationPolicy = ConfigurationPolicy.OPTIONAL)
+public class EnrouteApplicationIndexServer implements ConditionalServlet {
+
+	static final String NAME = "osgi.enroute.simple.web.application";
+
+	static class IndexDTO extends DTO {
```
Comment (pkriens) [inline osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/EnrouteApplicationIndexServer.java]: must be public

File: osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/EnrouteApplicationIndexServer.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+		List<ApplicationDTO>	applications	= new ArrayList<>();
+		Map<String,Object> configuration;
```
Comment (pkriens) [inline osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/EnrouteApplicationIndexServer.java]: DTO fields must also be public

File: osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/EnrouteApplicationIndexServer.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+		
+		static class ApplicationDTO extends DTO {
```
Comment (pkriens) [inline osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/EnrouteApplicationIndexServer.java]: must be public and probably better to move it one level up so it is a peer of the IndexDTO

**Comment by dleangen** [comment]
Thanks Peter. I have updated the PR as per your comments.

**Comment by pkriens** [comment]
The redirect servlet only redirects when the path ends with '/', ie. a directory. This fails the EnrouteApplicationIndexServer because it just uses the directory.

Obviously the easy fix is to change the html file and add /index.html to the link but I liked the model where you could just use the directory without the /, and I think this is the expected behavior for a lot of people?

File: osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/RedirectServlet.java [hunk 4]
```diff
@@ -53,22 +52,21 @@ public boolean doConditionalService(HttpServletRequest rq, HttpServletResponse r
 			String path = rq.getRequestURI();
 
 			if (path == null || path.isEmpty() || path.equals("/")) {
-				throw new RedirectException(redirect);
+				throw new Redirect302Exception(redirect);
 			} else if (path.startsWith("/"))
 				path = path.substring(1);
 
 			if (path.endsWith("/")) {
 				if (path.endsWith("/"))
```
Comment (pkriens) [inline osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/RedirectServlet.java]: this is always true ...

**Comment by dleangen** [comment]
I agree that this behaviour is expected for a lot of people. However, I don't think it's the correct behaviour. It is just a remnant from when a URI was essentially just mapped to a filesystem.

How about we make this behaviour configurable? We could even make the default such that it redirects, so that purists like myself would have to configure it not to redirect.

File: osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/RedirectServlet.java (hunk 4 above)
Comment (dleangen) [inline osgi.enroute.web.simple.provider/src/osgi/enroute/web/server/provider/RedirectServlet.java]: Updated. :-)
    Updated. :-)

**Comment by pkriens** [comment]
Can you make it configurable? (With the default being as it was, i.e. redirect on no slash). I guess the redirect is the last conditional servlet so that should be ok.

I think we can then merge!

**Comment by dleangen** [comment]
Actually, the redirect is the first servlet. Also, there is no way of knowing whether or not it is a directory unless you peer into the "file system" (in this case files in a bundle). So, it is quite a bit more complex than just determining whether or not the path ends with a "/".

Also, there are many possible representations for this case, your suggestion being one of them. Why only provide for this possibility?

Anyway, I have an idea as to how to handle this case. Working on it now...

**Comment by dleangen** [comment]
I am a bit stuck. Some help would be very much appreciated!! :-)

Because the workspace has changed, I need to re-resolve. The only change I can see to the runbundles in bnd.bnd (other than the osgi.enroute.web.simple.provider bundle that gets added in triplicate now), is that org.apache.felix.http.jetty gets updated from 3.1.0 to 3.2.0.

However, the tests no longer act as expected, and do not pass. The http requests are not getting routed to the correct servlet. A request to "http://localhost:8080/bnd/*" gets routed to the servlet having the context "/" (DispatchServlet), not the "/bnd" (BundleFileServer) servlet as expected.

I tried downloading the previous version of org.apache.felix.http.jetty and using that, but then the resolution fails at runtime.

If I run "gradle clean build", assuming the tests get run (which it looks like they don't), then build completes.

As a general comment, resolution does not seem to function like it used to, and I am getting a bit confused. I suppose that adding Maven support has added a bit of complications to the process, and some things are happening that I don't yet understand.

**Comment by dleangen** [comment]
I committed my changes anyway, even if I am unable to run the tests currently.

**Comment by dleangen** [comment]
Hi Peter,

I made the update, but could I ask you once again for a bit of help? There is one little thing left to be done.

You will notice that there should now be 7 test cases that fail. What I want to do is set the "addTrailingSlash" configuration parameter to "false", but only for those tests. Then these failing tests should pass again. However, I'm not sure how to set the config only for those test cases.

Sorry, should be obvious by now, I guess, but I'm not seeing it. :-(

**Comment by pkriens** [comment]
Can't you just set the configuration in Configuration Admin?

**Comment by dleangen** [comment]
I did try, but there is some kind of "cross-over" with the other tests. As soon as I do that, a whole bunch of other tests fail (when normally there should be no effect whatsoever).

For now, I will comment out the failing tests so at least we can merge these changes and work from a smaller PR.

**Comment by dleangen** [comment]
Ok, I commented out the failing tests. I will work on those in a separate PR.

Trailing slashes have been reverted to previous behaviour as default, so I think we're good to go!

**Comment by dleangen** [comment]
Failing tests are from the Scheduler. Not sure why. They are passing locally.

**Comment by dleangen** [comment]
Scheduler tests are passing now. I don't know why: I didn't change anything!

Good to merge, if you accept.

**Pull request merged**